hack/functests.sh

```diff
@@ -37,4 +37,10 @@ if [[ ${KUBEVIRT_PROVIDER} == os-* ]] || [[ ${KUBEVIRT_PROVIDER} =~ (okd|ocp)-*
     oc=${kubectl}
 fi
 
-${TESTS_OUT_DIR}/tests.test -kubeconfig=${kubeconfig} -container-tag=${docker_tag} -container-tag-alt=${docker_tag_alt} -container-prefix=${functest_docker_prefix} -image-prefix-alt=${image_prefix_alt} -oc-path=${oc} -kubectl-path=${kubectl} -gocli-path=${gocli} -test.timeout 420m ${FUNC_TEST_ARGS} -installed-namespace=${namespace} -previous-release-tag=${PREVIOUS_RELEASE_TAG} -previous-release-registry=${previous_release_registry} -deploy-testing-infra=${deploy_testing_infra}
+test_options="-kubeconfig=${kubeconfig} -container-tag=${docker_tag} -container-tag-alt=${docker_tag_alt} -container-prefix=${functest_docker_prefix} -image-prefix-alt=${image_prefix_alt} -oc-path=${oc} -kubectl-path=${kubectl} -gocli-path=${gocli} -test.timeout 420m ${FUNC_TEST_ARGS} -installed-namespace=${namespace} -previous-release-tag=${PREVIOUS_RELEASE_TAG} -previous-release-registry=${previous_release_registry} -deploy-testing-infra=${deploy_testing_infra}"
+if [[ ${KUBEVIRT_PROVIDER} =~ .*(k8s-1\.16)|(k8s-1\.17)|k8s-sriov.* ]]; then
+    echo "Will skip test asserting the cluster is in dual-stack mode."
+    test_options="$test_options -skip-dual-stack-test"
```

Comment:
Can you explain why we need this extra flag? If I see IPv6 IPs reported we can assume that they just work. This introduces a third way on how to focus/skip on tests:

 * selecting tagged tests in `automation/tests.sh`
 * Dynamic detection

And now this extra flag which I may have to maniually add or remove.

Please consolidate this. 

I would assume that if IPv6 IPs are reported on Pods that they are just supposed to work and they should be tested. If you fear that you may miss running IPv6/dual-stack tests in a cluster which should be dual-stack but is in fact only ipv4, consider using tags and focus on them, instead of the flag. Otherwise it is fine to assume that any IP which is reported on nodes and pods should just work in clusters and we can assume ipv6 support on clusters where pods report ipv6 addresses. In general we do a lot of auto-discovery already. Yes we can miss a test, but it makes integration very easy and we could think about pre-flight checks to see if the cluster is in the shape which we want it to be. The applied regex solution which we have just moves the do-we-really-run-the-tests question to the next auto-discovery based on regex instead of cluster-introspection.

Reply:
> Can you explain why we need this extra flag? If I see IPv6 IPs reported we can assume that they just work. This introduces a third way on how to focus/skip on tests:

We're currently converting the (most representative) existing networking tests to run in dual stack (when lanes support it).

We've chosen an approach where we don't check the type of cluster in the test itself, we just iterate the IPs, and act accordingly (create one service per IP address on the primary pod interface in the services test is an example).

Since we're just iterating it, we could reach a scenario where the tests pass but the cluster is (for whatever reason) not in dual stack mode, giving us a false positive. This way, we run an extra test (only on lanes that support dual-stack) that actively checks if the cluster networking configuration is accurate.

> 
> * selecting tagged tests in `automation/tests.sh`
> * Dynamic detection
> 
> And now this extra flag which I may have to maniually add or remove.
> 
> Please consolidate this.
> 
> I would assume that if IPv6 IPs are reported on Pods that they are just supposed to work and they should be tested. If you fear that you may miss running IPv6/dual-stack tests in a cluster which should be dual-stack but is in fact only ipv4, consider using tags and focus on them, instead of the flag. 

I can adjust this on a follow up PR to work like this (add a tag to the added test, then include that test in the focus list of the lanes) if you really want.

> Otherwise it is fine to assume that any IP which is reported on nodes and pods should just work in clusters and we can assume ipv6 support on clusters where pods report ipv6 addresses. In general we do a lot of auto-discovery already. Yes we can miss a test, but it makes integration very easy and we could think about pre-flight checks to see if the cluster is in the shape which we want it to be. 

That's what we don't want, under any circumstance: miss a test.

I could also convert this test into a pre-flight check, but I think the timing for it would be only when all supported lanes run on dual-stack clusters (supported lanes would be k8s-1.18, k8s-1.19, k8s-1.20) .

> The applied regex solution which we have just moves the do-we-really-run-the-tests question to the next auto-discovery based on regex instead of cluster-introspection.

@EdDev : do you have anything to add to this discussion ?

Reply:
> That's what we don't want, under any circumstance: miss a test.

This can happen right now in many cases. We can definitely improve it, but having such non-holistic solutions just complicate everything immensly. This is not a solution, because we would have to add skip-flags for a vast amount of scenarios.

One solution would for instance be explicitly listing up-front which tests are supposed to be skipped and then comparing that list with the resulting junit result. This island-solution just complicates things without solving anything as a whole.

Reply:
There aren't any tests to be skipped; the tests are the same under ipv4 or dual stack. They will just test one or multiple IP families.

Reply:
> There aren't any tests to be skipped; the tests are the same under ipv4 or dual stack. They will just test one or multiple IP families.

Well it is an implicit skip. I would suggest to separate the tests and ensure that the tests have a clear scenario which is focused on the cluster capabilities. you can then also tag them accordingly. This was a little bit hard to achieve with IPv6 only stacks, but now with dual stack it is easy to do. It has also much more meaning if I see in my output ipv4-network-test passed, ipv6-network-test failed, vs. "some IPs could not be reached".

Reply:
I - and I think the whole networking team - don't think so. We want to have a single test to check both families, thus avoiding reducing a ton of code. Which will make everything harder to understand, and especially maintain.

I don't think your argument about meaning above is valid; "ipv4-network-test passed; ipv6-network-test failed" has the exact same meaning of "network test failed .... Succeeded to reach over IPv4 addr; Failed to reach over IPv6 address".

Anyway, I see there are 2 discussions happening at once here:
  - the flag
  - testing scenarios, instead of scenarios per IP family.

I'm willing to come up with an alternative solution for the flag (add the focus to the lanes on `hack/functest.sh`, as done for the others), if you want, but duplicating the test cases is something we as a team went over and discarded, for the reasons explained above.

Reply:
> I - and I think the whole networking team - don't think so. We want to have a single test to check both families, thus avoiding reducing a ton of code. Which will make everything harder to understand, and especially maintain.

I kindly disagree and you are adding implicit skip mechanisms where the normal skip mechanisms are good enough. It is definitely harder to see on the test result what was tested. It looks to me like you are working against the test framework and we would have many scenarios where we could do implicit checks and test more-or-less, depending on cluster capabilities or states. It is just very hard to maintain and see what is going on this way.

As said to guarantee that tests get executed, I am strongly favouring comparing expected tests with executed tests.

Reply:
- Flag: The current filtering is performed against the kubevirtci lanes, aimed at our own CI run. The solution needs to be cross target clusters, including external clusters (e.g. openshift) and serving things like conformance tests.
In addition, when running locally the CI script is not executed.

Bottom line, we need an integral cross cluster solution and adding configuration to a test binary is already provided using flags (there are a lot). We could load a configuration yaml in as an alternative, of build it with focuses/skips as suggested, but then it cannot live in the current script.
IMO the focus/skip mechanism is messy, I like it less and I am unsure if one can provide concatenated filters (e.g. using the same flag multiple times, each added as a "or" or "and".

- Dual Stack tests: We usually test a feature/scenario. The fact that the same scenario may come with a parameterization which is cross all network tests, means that we either:
   - have to build two lanes.
   - have to duplicate the number of tests, implying double the time, increase code extensively and duplication.
   - add to existing tests another check.

We choose the 3rd option, as it is much quicker, easy to read and is not dependent on anything except what the setup provides (i.e. no conditions, not special cases). It works for single stack IPv4, single stack IPv6 and dual stack.
It will/should work if the order of the IP/s in the list will change (currently IPv4 is always first) or when k8s will realize that the Pod may have multiple IP/s on the same version (especially with IPv6 which is very common).

From a run point of view, there is no need to create and destroy VMI/s or other resources for each address, we can keep the tests self contained and avoid leaking resources to be left to the next test (currently handled by the "big" cleanup of the namespace).

We skip nothing, we just follow what the cluster gives us and handle it in a generic clean manner.
We have identified the problem that we may have in our hand a false-positive if the tester executes the tests on a single stack cluster and he on the other hand thinks he is checking dual-stack. For that an explicit test was created and discussed in the previous section.

> As said to guarantee that tests get executed, I am strongly favouring comparing expected tests with executed tests.

This will cause a maintenance cost which I do not think anyone will be willing to commit to. Tests may move, names may change, one needs to know that if one test is added, it needs to be added in some other list. Same for removals.
I do not see why we should pay such a penalty.

Reply:
> * Flag: The current filtering is performed against the kubevirtci lanes, aimed at our own CI run. The solution needs to be cross target clusters, including external clusters (e.g. openshift) and serving things like conformance tests.
>   In addition, when running locally the CI script is not executed.

Autodiscovery can do that too. You just need to check the resulting junit xml to see if tests were executed.

> We choose the 3rd option, as it is much quicker, easy to read and is not dependent on anything except what the setup provides (i.e. no conditions, not special cases). It works for single stack IPv4, single stack IPv6 and dual stack.
> It will/should work if the order of the IP/s in the list will change (currently IPv4 is always first) or when k8s will realize that the Pod may have multiple IP/s on the same version (especially with IPv6 which is very common).

I don't understand what is so complicated on a table test with two entries where the last part of the test is `for ipv6 addresses` on one entry and `for ipv4 addresses` on the other one. The test can have the same body, you basically just filter out ipv4 and ipv6 addresses.

I am not saying that I always prefer full recycling but whenever possible (like when using small VMs) it should be the default decision. We had so many issues due to shared resources between tests or tests scenarios, especially in the networking tests. #4051 for parallel executions is close to be merged, to speed the tests up. It saves much more time than all kinds of implicit individual-test-level improvements can do.

> This will cause a maintenance cost which I do not think anyone will be willing to commit to. Tests may move, names may change, one needs to know that if one test is added, it needs to be added in some other list. Same for removals.
> I do not see why we should pay such a penalty.

Yes it would cause maintenance but so does yours, my point is that in this case with your flag you solve something for your small scenario and force every executor of the test to learn about the new flags, learn about the implications of this flag, and since you see this as the preferred pattern you would want that others add such skip flags for all other kinds of issues. I don't see how this scales better.

There are a huge amount of scenarios where tests may be skipped if something is wrong with the cluster. One would not really want to add extra flags for all these and maintain these. It is at the end a decision which the project needs to do, if it wants better checking if the expected tests were executed, but I don't see why verifying a small sub-test in a huge suite as such adds value compared to the complexity.

> We skip nothing, we just follow what the cluster gives us and handle it in a generic clean manner.

I don't see how skipping parts of a test is not skipping. Having a cluster with IPv6 support is definitely a huge feature in k8s, so having sub-scenarios to explicitly express the support for it makes perfectly sense. For instance without looking and closely studying your tests I will have no clue, if we support IPv6 connections to VMs. I could run the tests on a dual stack cluster, the test would pass and it could as well just mean: Yes kubevirt does not crash if you have dual stack enabled.

> IMO the focus/skip mechanism is messy, I like it less and I am unsure if one can provide concatenated filters (e.g. using the same flag multiple times, each added as a "or" or "and".

Yes it is not perfect and that is not new (just look up the issues on the ginkgo framework where the k8s people are reporting that), but it is what we have right now and it works, adding flags all over the place does not help.

Reply:
What if this added test is ignored if the flag is not supplied ? That would remove the non-holistic skip based approach.

Would you accept this compromise ?